v0.30: cancel tasks (#1981)
meilisearch/documentation: reference/api/tasks.md

#### `taskCancelation`

| Name                  | Description                                                                                                                                                                                          |
| :-------------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | **`matchedTasks`**    | The number of tasks that can be canceled based on the request. If the API key doesn’t have access to any of the indexes specified in the request, those tasks will not be included in `matchedTasks` |

Reviewer: See https://github.com/meilisearch/documentation/pull/1983#discussion_r1021578395

| **`canceledTasks`**   | The number of tasks successfully canceled. If the task fails, this will be `0`                                                                                                                       |
| **`originalFilters`** | The filter used in the [`/tasks/cancel`](#cancel-tasks) request                                                                                                                                      |

#### `taskDeletion`

### `error`

**Type**: Object

> **Description**: Error details and context. Only present when a task has the `failed` [status](#status)

Reviewer 2: ```suggestion
**Description**: Error details and context. Contains the error definition, if this task has `failed`. [status](#status). Otherwise, set to `null`
```

> **Description**: Error details and context. Contains the error definition, if this task has `failed` [status](#status). Otherwise, set to `null`

Reviewer 3: ```suggestion
**Description**: If the task has `failed` [status](#status), contains the error definition. Otherwise, set to `null`
```

## Cancel tasks

<RouteHighlighter method="POST" route="/tasks/cancel?{task_uid_or_status_or_type_indexUid_or_afterXAt_or_beforeXAt}"/>

> Cancel an `enqueued` or `processing` task based on `uid`, `status`, `type`, `indexUid`, and date. Task cancelation is an atomic transaction, **either all tasks are successfully canceled or none are**.

Author: ```suggestion
Cancel an `enqueued` or `processing` task based on `uid`, `status`, `type`, `indexUid`, or date. Task cancelation is an atomic transaction, **either all tasks are successfully canceled or none are**.
```

::: warning
Using this route without any filters (POST `/tasks/cancel`) will result in the [`missing_filters`](/reference/errors/error_codes.md#missing-filters) error.
:::

The API key used must have the `tasks.cancel` action.

You can also cancel `taskCancelation` type tasks as long as they are in the `enqueued` or `processing` state.

### Query parameters

> A valid `uid`, `status`, `type`, `indexUid`, and date(`beforeXAt` or `afterXAt`) is required.

Author: ```suggestion
A valid `uid`, `status`, `type`, `indexUid`, or date(`beforeXAt` or `afterXAt`) is required.
```

| **`beforeFinishedAt`** | Delete tasks before a specified `finishedAt` date                                                                                    |
| **`afterEnqueuedAt`**  | Delete tasks after a specified `enqueuedAt` date                                                                                     |
| **`afterStartedAt`**   | Delete tasks after a specified `startedAt` date                                                                                      |
| **`afterFinishedAt`**  | Delete tasks after a specified `finishedAt` date                                                                                     |

### Example

<CodeSamples id="cancel_task_1" />

#### Response: `200 Ok`

```json
{
  "taskUid": 3,
  "indexUid": null,
  "status": "enqueued",
  "type": "taskCancelation",
  "enqueuedAt": "2021-08-12T10:00:00.000000Z"
}
```
::: note
Since `taskCancelation` is a [global task](/learn/advanced/asynchronous_operations.md#global-tasks), the `indexUid` is `null`.
:::

You can use this `taskUid` to get more details on the [status of the task](#get-one-task).

### Cancel all tasks

You can cancel all tasks by using the following filter:

> <RouteHighlighter method="POST" route="/tasks/cancel?status=processing,enqueued" />

Reviewer 2: ```suggestion
<RouteHighlighter method="POST" route="/tasks/cancel?statuses=processing,enqueued" />
```

## Cancel tasks

<RouteHighlighter method="POST" route="/tasks/cancel?{task_uid}"/>

> Cancel an `enqueued` or `processing` task based on `uid`, `status`, `type`, `indexUid`, or date. Task cancelation is an atomic transaction, **either all tasks are successfully canceled or none are**.

Reviewer 3: I recommend a longer pause here because it gives a stronger sense of explanation/conclusion.

```suggestion
Cancel an `enqueued` or `processing` task based on `uid`, `status`, `type`, `indexUid`, or date. Task cancelation is an atomic transaction: **either all tasks are successfully canceled or none are**.
```

::: warning

> Using this route without any filters (POST `/tasks/cancel`) will result in the `missing_task_filters` error. This error prevents users from accidentally canceling all `enqueued` and `processing` tasks.

Reviewer 3: Might be personal preference, but I would have gone the other way around with this note. That is, "to prevent users from accidentally cancelling all tasks, Meilisearch throws an error" instead of "the error prevents users from accidentally cancelling all tasks." Like, emphasise the reason why we don't want people to use the route without filters instead of the error itself.

| **`indexUids`**        | Delete tasks based on `indexUid`. Separate multiple `uids` with a comma (`,`). Use `indexUids=*` for all `indexUids`. Case-sensitive |
| **`beforeEnqueuedAt`** | Delete tasks **before** a specified `enqueuedAt` date                                                                                    |
| **`beforeStartedAt`**  | Delete tasks **before** a specified `startedAt` date                                                                                     |
| **`beforeFinishedAt`** | Delete tasks **before** a specified `finishedAt` date                                                                                    |
| **`afterEnqueuedAt`**  | Delete tasks **after** a specified `enqueuedAt` date                                                                                     |
| **`afterStartedAt`**   | Delete tasks **after** a specified `startedAt` date                                                                                      |
| **`afterFinishedAt`**  | Delete tasks **after** a specified `finishedAt` date                                                                                     |

::: note

> The date filters are exclusive, meaning you can only filter tasks before or after a specified date.

Reviewer 3: ```suggestion
Date filters are exclusive, meaning you can only filter tasks before or after a specified date.
```

:::

### Example

<CodeSamples id="cancel_task_1" />

#### Response: `200 Ok`

```json
{
  "taskUid": 3,
  "indexUid": null,
  "status": "enqueued",
  "type": "taskCancelation",
  "enqueuedAt": "2021-08-12T10:00:00.000000Z"
}
```

::: note

> Since `taskCancelation` is a [global task](/learn/advanced/asynchronous_operations.md#global-tasks), the `indexUid` is `null`.

Reviewer 3: ```suggestion
Since `taskCancelation` is a [global task](/learn/advanced/asynchronous_operations.md#global-tasks), its `indexUid` is `null`.
```
